File: playwright-java/src/main/java/testtracks/cleanup/CleanupRegistry.java
package testtracks.cleanup;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public final class CleanupRegistry {
  private final Stack<CleanupAction> actions = new Stack<>();

  public void register(String description, CleanupActionBody action) {
    actions.push(new CleanupAction(description, action));
  }

  public List<Exception> run() {
    var failures = new ArrayList<Exception>();

    while (!actions.empty()) {
      var cleanup = actions.pop();
      try {
        cleanup.action().run();
      } catch (Exception ex) {
        failures.add(new IllegalStateException("Cleanup failed: " + cleanup.description(), ex));
      }
    }

    return failures;
  }

  @FunctionalInterface
  public interface CleanupActionBody {
    void run() throws Exception;
  }

  private record CleanupAction(String description, CleanupActionBody action) {}
}

File: playwright-csharp/src/Cleanup/CleanupRegistry.cs
namespace TestTracks.Playwright.CSharp.Cleanup;

/// <summary>
/// Stores cleanup actions for data created by a scenario.
/// </summary>
public sealed class CleanupRegistry
{
    private readonly Stack<CleanupAction> _actions = new();

    /// <summary>
    /// Registers a cleanup action to run after the scenario.
    /// </summary>
    /// <remarks>
    /// Cleanups run in reverse order, which usually matches how dependent test data should be removed.
    /// </remarks>
    public void Register(string description, Func<Task> action)
    {
        _actions.Push(new CleanupAction(description, action));
    }

    /// <summary>
    /// Runs all registered cleanup actions and returns any failures without stopping at the first one.
    /// </summary>
    public async Task<IReadOnlyList<Exception>> RunAsync()
    {
        var failures = new List<Exception>();

        while (_actions.TryPop(out var cleanup))
        {
            try
            {
                await cleanup.Action();
            }
            catch (Exception ex)
            {
                failures.Add(new InvalidOperationException(
                    $"Cleanup failed: {cleanup.Description}", ex));
            }
        }

        return failures;
    }

    private sealed record CleanupAction(string Description, Func<Task> Action);
}
